### PR TITLE
Client Patching (for 2.8 release)

### PR DESCRIPTION
**Pull request by 4Benj**
~~This was developed during the beta and has not yet been tested for the preload (Still waiting to finish the download lol).
I recommend making a new branch for this for a bit of clean up.~~
This has been verified as working with the preload.

**If you patch your game using the proxy make sure you uncheck "Connect via Grasscutter" and start the game to restore your game to an unpatched state before using the official launcher and connecting to official servers.**
**Use at your own risk**

A major bug I've noticed:
- Patching takes too long and the proxy starts up before the game launches, so the proxy closes as it doesn't detect the game. In order to fix this, just close the game and relaunch with clicking anything else. Since the game is already patched, the game launches faster and the proxy detects the game.

### CODE REVIEW

**Comment by MoonLGH** [comment]
okay lets start with, why `https://github.com/Grasscutter-Backrooms/Cultivation/blob/patching/package.json` is still on 1.0.1

**Comment by 4Benj** [comment]
> okay lets start with, why [version](https://github.com/Grasscutter-Backrooms/Cultivation/blob/patching/package.json) is still on 1.0.1

I actually got this working a couple days ago and I just forgot to update to the latest commit before uploading here to GitHub. The PR is still a draft for a couple reasons, one of them being that I haven't updated it to the latest commit of cultivation (although I doubt there would be any major issues)

File: src/ui/components/menu/Options.tsx [hunk 1]
```diff
@@ -36,6 +37,7 @@ export default class Options extends React.Component<IProps, IState> {
 
     this.state = {
       game_install_path: '',
+      game_executable: '',
```
Comment (SpikeHD) [inline src/ui/components/menu/Options.tsx:40]: Again, there should only need to be one option for this. If you want to put functions that get the game folder/exe name/metadata path inside of their own file in `src/utils` you are more than welcome to

File: src/ui/components/menu/Options.tsx [hunk 2]
```diff
@@ -58,49 +61,68 @@ export default class Options extends React.Component<IProps, IState> {
...
+  setGamePath(value: string) {
     setConfigOption('game_install_path', value)
 
     this.setState({
-      game_install_path: value
+      game_install_path: value,
     })
   }
 
+  setGameExecutable(value: string) {
```
Comment (SpikeHD) [inline src/ui/components/menu/Options.tsx:95]: A less vital spot but early returns here would be nice too (and if there are parts of the code where I have not done that dw I will go through and fix em all up to be uniform at some point)

File: src-tauri/keys/passwordKey.txt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<RSAKeyValue>
```
Comment (SpikeHD) [inline src-tauri/keys/passwordKey.txt:1]: Is it safe to just keep the key here?

**Comment by SpikeHD** [comment]
Oh, yeah, one final thing I meant to mention, Cultivation should always replace the patched metadata with the original upon closing the game, in the case of people who still use the official launcher to play the official game.

Should be a trivial addition, you can see where Culti does this by taking a look at [src-tauri/src/main.rs](https://github.com/Grasscutters/Cultivation/blob/4fc90ee333fefa8a8e60bab02e7dee648acf7700/src-tauri/src/main.rs#L81), where you should just be able to pop a function call in there (and yes, I know what exists is a lil bit of a scuffed solution but I intend to improve it sometime in the future)

**Comment by KingRainbow44** [comment]
> Patching takes too long and the proxy starts up before the game launches

any status update on this? to me it sounds like we might need to rewrite the game watcher

**Comment by 4Benj** [comment]
> The problem is weird tho because all metadata related function calls are awaited before the proxy and game are even meant to start so idrk whats up with it.

Maybe I can do stuff with Promises as a temporary fix? awaits rarely seem to work properly for me in js lol.

File: src/ui/components/ServerLaunchSection.tsx [hunk 4]
```diff
@@ -85,19 +107,103 @@ export default class ServerLaunchSection extends React.Component<IProps, IState>
...
+      return false
+    }
+  }
+
   async playGame() {
     const config = await getConfig()
-  
-    if (!config.game_install_path) return alert('Game path not set!')
+
+    if(this.getGameExecutable(config) == null) {
```
[the remark this answers is not in the document]
Comment (SpikeHD) [reply src/ui/components/ServerLaunchSection.tsx:151]: also prefer `===` over `==` whenever possible for strict type equality.

[the remark this answers is not in the document]
Comment (4Benj) [reply src/ui/components/ServerLaunchSection.tsx:151]: My bad, old habits lol

**Comment by 4Benj** [comment]
Closing since this is being merged into a special patch branch

**Pull request closed without merging**